File: 0115-distinct-subsequences/0115-distinct-subsequences.py
class Solution:
    def numDistinct(self, s: str, t: str) -> int:
        cache = {};
        def rec(i,j):
            if(j < 0): return 1;
            if(i < 0): return 0;
            
            if (i,j) in cache:
                return cache[(i,j)];
            if(s[i] == t[j]):
                cache[(i,j)] = rec(i-1,j-1) + rec(i-1,j);
            else: 
                cache[(i,j)] = rec(i-1,j);
            return cache[(i,j)];
        return rec(len(s)-1,len(t)-1);

File: 0115-distinct-subsequences/0115-distinct-subsequences.cpp
class Solution {
public:
    int rec(int i, int j, string a, string b, vector<vector<int>> &dp){
        if(j < 0) return 1;
        if(i < 0) return 0;

        if(dp[i][j] != -1) return dp[i][j];
        if(a[i] == b[j]){
            return dp[i][j] = rec(i-1,j-1,a,b,dp) + rec(i-1,j,a,b,dp);
        }

        return dp[i][j] = rec(i-1,j,a,b,dp);
    }
    int numDistinct(string s, string t) {
        vector<vector<int>> dp(s.size(),vector<int>(t.size(),-1));
        return rec(s.size()-1,t.size()-1,s,t,dp);
    }
};